File: Problems/Algorithms/704. Binary Search/binary_search.rs
use std::cmp::Ordering;

impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let (mut left, mut right) = (0, nums.len());
        
        while left < right {
            let mid = left + (right - left) / 2;
            
            match nums[mid].cmp(&target) {
                Ordering::Equal => return mid as i32,
                Ordering::Greater => right = mid,
                Ordering::Less => left = mid + 1,
            }
        }
        
        -1
    }
}

File: Problems/Algorithms/704. Binary Search/binary_search.go
package main

func search(nums []int, target int) int {
	left, right := -1, len(nums)

	for right-left > 1 {
		mid := left + (right-left)/2
		if nums[mid] == target {
			return mid
		} else if nums[mid] < target {
			left = mid
		} else {
			right = mid
		}
	}
	return -1
}
